### ISSUE TITLE
[Bug]: speculative decoding under pipeline parallelism produces wrong output with --no-async-scheduling

### ISSUE THREAD

**Issue by nickus** [vllm-project/vllm#52071]
### Your current environment

vLLM `0.26.1rc1.dev693+g7f7a32cfe` (nightly), 8× RTX 3090 (sm_86), CUDA 12.x,
`VLLM_USE_V2_MODEL_RUNNER=1`. Reproduced at `pipeline_parallel_size` 2, 4 and 8,
with two different speculative methods and two model families. Code references
are against `main` @ `98f86b9c0232`.

### 🐛 Describe the bug

With `--no-async-scheduling`, speculative decoding under pipeline parallelism
produces wrong output — or a CUDA fault — instead of being rejected or working.
Turning async scheduling back on fixes it.

### Root cause

`next_decode_eligible_step` is assigned in exactly one place in the tree:

```
vllm/v1/core/sched/async_scheduler.py:49
    request.next_decode_eligible_step = self.current_step + self.pp_size
```

initialised in one place:

```
vllm/v1/request.py:167
    self.next_decode_eligible_step = 0
```

and read in exactly one place:

```
vllm/v1/core/sched/scheduler.py:509
    if self.current_step < request.next_decode_eligible_step:
        # V2+PP+async: enforce `pp_size` steps between same-req decodes
        # to match worker-side sampled-tokens broadcast slot ring cadence.
```

The base `Scheduler` — which is what `--no-async-scheduling` selects — reads that
field and never sets it. It stays `0`, the guard never fires, and nothing keeps a
request's decodes `pp_size` steps apart. The worker-side sampled-token broadcast
ring is then read out of phase.

### What that produces, measured

A probe at the point where a step's input ids are built
(`combine_sampled_and_draft_tokens`), PP=2:

```
last_sampled=[16925]  drafts=[16925, 16925, 16925]    healthy
last_sampled=[0]      drafts=[0, 0, 0]                ~25% of spec steps
```

Token id 0 is embedded as the anchor. On the model under test token 0 is `!`, and
the generated text fills with `!`. The same skew has a second face — the step gets
scheduled without the accepted token that its drafts hang off:

```
healthy:  hidden=(4,1024)  logits_indices=[0, 1, 2, 3]   qlen=4  num_logits=4  ndraft=3
failing:  hidden=(3,1024)  logits_indices=[-1, 0, 1, 2]  qlen=3  num_logits=4  ndraft=3
```

`logits_start = query_end - num_logits = 3 - 4 = -1`, so
`hidden_states[input_batch.logits_indices]` either wraps to the last row silently
(torch semantics) or trips:

```
Assertion `ind >=0 && ind < ind_dim_size && "vectorized gather kernel index out of bounds"` failed
```

The same skew also reaches `_prepare_prefill_inputs_kernel` in
`v1/worker/gpu/spec_decode/autoregressive/speculator.py`, where
`query_len -= num_rejected` can reach 0 and the unmasked
`tl.store(draft_input_ids_ptr + query_start - 1, ...)` writes out of bounds —
Triton adds negative indices to the base pointer rather than wrapping.

### Reproduced with

- `method: mtp` on GLM-4.5-Air (AWQ), PP=4
- `method: dspark` on Kimi-K3, PP=8
- PP=2 on a 16-layer synthetic target

### Controls

Two controls that rule out the obvious alternatives, same tree, same session:

```
speculation OFF, async on vs async off   : outputs IDENTICAL
Qwen3-0.6B, PP=1 vs PP=2, no speculation : outputs IDENTICAL
```

So the pipeline path is greedy-stable on its own, and the scheduler mode alone
changes nothing. The defect appears only when speculation and PP are combined
with async scheduling disabled.

Also worth noting: with async scheduling ON, plain decode throughput on the same
config was ~1.9× better than with it off (94.6 → 48.8 ms/token TPOT, GLM-4.5-Air
PP=4, concurrency 1), so the flag is expensive in its own right.

### Suggested fix

Any of these beats the current behaviour, in descending order of preference:

1. Set the cadence in the base scheduler too, so the guard works regardless of
   scheduling mode.
2. Reject `speculative_config` + `pipeline_parallel_size > 1` +
   `--no-async-scheduling` at config time, with a message naming the cadence.
3. At minimum, document that speculative decoding under PP requires async
   scheduling.

Silently emitting `!` is the worst of the available outcomes, and it is what
happens today.

### Before submitting a new issue...

- [x] I searched the tracker (several phrasings) and did not find this reported.

---
Assisted-By: Claude

**Comment by aoshen02** [comment]
Could you please raise a PR?

**Comment by lunarlinger** [comment]
Hi, I'd like to work on this. I'll investigate the scheduler cadence handling and try to add a fix plus regression tests for the --no-async-scheduling path.

**Comment by nickus** [comment]
@aoshen02 Yes — the fix for #52069 (the config-time gate that sits in front of this) is up as #52117. For this issue the root cause is pinned and the fix is small; validation of the exact hunk is still in progress, details below.

@icarus0adios-netizen glad to have you on it — here is everything we have, so nothing gets re-derived:

**Root cause (measured, not inferred).** `next_decode_eligible_step` is assigned only in `async_scheduler.py:49` and read only at `scheduler.py:509`; the base scheduler never sets it, so with `--no-async-scheduling` the guard never fires and a spec step can execute before its anchor token returns over the deferred PP relay. Probed at input-build time: `last_sampled=[0] drafts=[0,0,0]` on ~25% of spec steps, and the model embeds token id 0.

**Proposed fix** (branch: https://github.com/nickus/vllm/tree/fix-pp-spec-cadence): mirror the assignment into the base scheduler's post-schedule update, gated on `speculative_config is not None and pipeline_parallel_size > 1` — without speculation the base scheduler is already correct under PP (greedy output byte-identical PP=1 vs PP=2, measured).

**Status of validation:** the diagnosis ran on 0.26.1 nightlies (three machines, MTP on GLM-4.5-Air and DSpark on Kimi-K3); the hunk itself I have not yet exercised end-to-end — on today's `0.27.2rc1.dev5` my 2-GPU repro stand hit an unrelated-looking engine death at PP=1 that I did not have rented-GPU budget left to chase.

If you want to fold this hunk into your PR together with the regression tests you mentioned, please do — happy to review. Otherwise I will finish validation and open the PR myself in the next day or two. A useful regression shape either way: same prompt, greedy, spec on, `--no-async-scheduling`, assert PP=2 output equals PP=1 — that is the exact check that fails today.

---
Assisted-By: Claude

**Comment by lunarlinger** [comment]
Thanks @nickus for the detailed reproduction and root-cause analysis, especially the sampled-token relay cadence observations — they made it much easier to narrow the fix down to the scheduler side.

I've implemented a fix that sets `next_decode_eligible_step` in the base scheduler for V2 + PP + speculative decoding when async scheduling is disabled, while keeping chunked prefill unaffected.

I also added a scheduler regression test and validated the fix with a 2× RTX 4090 GPU E2E setup.

Results so far:

- Targeted scheduler regression: passed
- `test_async_scheduler.py`: 16 passed
- `test_scheduler.py -k spec`: 17 passed
- GPU E2E, V2 + MTP + `async_scheduling=False`:
  - before fix: PP=1 passes, PP=2 reproduces a CUDA illegal memory access
  - after fix: PP=1 and PP=2 both pass
  - PP=1 / PP=2 produce identical 16-token outputs
  - speculative decoding is active (`15` draft steps)

I'm going to open a PR with the fix and regression coverage shortly.

**Comment by nickus** [comment]
Excellent — that matches the fix shape exactly, and the 2×4090 E2E result mirrors what we measured. Happy to also validate your PR on 8× RTX 3090 at PP=8 with a DSpark draft on a Kimi-K3 slice once it's up — that covers the second speculative method and a deeper pipeline. Thanks for picking it up so quickly.

---
Assisted-By: Claude

**Comment by lunarlinger** [comment]
Thanks! The PR is up now. PP=8 + DSpark would be really helpful to validate the fix on a deeper pipeline.
